apache2/apache2_util.c

```diff
@@ -187,6 +189,83 @@ char *get_env_var(request_rec *r, char *name) {
     return result;
 }
 
+/**
+ * Retrieve waf log field.
+ */
+void get_field_value(const char* from, const char* to, const char* text, char* output) {
+    char* first = strstr(text, from);
+    int first_index = first - text;
+
+    if (first != NULL ) {
+        if ((first_index > 0 && (first-1)[0] != "\\") || (first_index == 0)) {
+            first += strlen(from);
+        }
+        else {
+            first = NULL;
+        }
+    }
+
+    char* last = strstr(first, to);
+    int last_index = last- first;
+    if (last != NULL ) {
+        if ((last_index > 0 && (last-1)[0] != "\\") || (last_index == 0)) {
+        }
+        else {
+            last = NULL;
+        }
+    }
+
+    if (first != NULL && last != NULL) {
+        strncpy(output, first, last - first);
+    }
+}
+
+/**
+ * send all waf fields in json format to a file.
+ */
+void send_waf_log(const char* str1, const char* ip_port, const char* uri, int mode, const char* hostname, request_rec *r) {
+    int rc = 0;
+    char waf_filename[1024] = "";
+    char waf_line[1024] = "";
+    char waf_id[1024] = "";
+    char waf_message[1024] = "";
+    char waf_data[1024] = "";
+    char waf_ip[50] = "";
+    char waf_port[50] = "";
+    char waf_detail_message[1024] = "";
+
+    get_field_value("[file ", "]", str1, waf_filename);
+    get_field_value("[id ", "]", str1, waf_id);
+    get_field_value("[line ", "]", str1, waf_line);
+    get_field_value("[msg ", "]", str1, waf_message);
+    get_field_value("[data ", "]", str1, waf_data);
+
+    char *comma = strstr(ip_port, ":");
+    if (comma!= NULL) {
+        strcpy(waf_port, comma+1);
+        strncpy(waf_ip, ip_port, comma - ip_port);
+    }
+    else {
+        strcpy(waf_ip, ip_port);
+    }
+
+    char *end = strstr(str1, "[file ");
+    if (end != NULL) {
+        strncpy(waf_detail_message, str1, end - str1);
+    }
+
+    rc = write_json_to_file(waf_ip, waf_port, uri, "", "", waf_id, waf_message, mode, 0, waf_detail_message, waf_data, waf_filename, waf_line, hostname);
```

Comment:
"" or NULL?

Reply:
will add directory to set rule set and version